### DISCUSSION TITLE
[테스트 셋업 길드]AnalyticsProviderContext 테스트 코드 관련해서 여쭤보고싶어요 ~!

### DISCUSSION THREAD

**Discussion by josworks27** [EveryAnalytics/react-analytics-provider#128] [Ideas]
# 개요
AnalyticsProviderContext에 대해 테스트 코드를 작성해 보려고 하는데, context api의 테스트 코드가 익숙하지 못해 어떤 식으로 접근하면 좋을지 논의해 보고 싶어요

# 궁금점
구글링해서 보면 대체로 context의 값을 컨슈머 컴포넌트에서 렌더링 시켜서 해당 값이 잘 렌더링 되었는지 확인하는 테스트 코드 위주가 많습니다.
그게 아닌, 저희 처럼 이벤트 전송 함수를 어떻게 테스트 해보면 좋을 지 조언 부탁드리고 싶습니다 !

# 참고한 자료
* https://testing-library.com/docs/example-react-context/
* https://www.samdawson.dev/article/react-context-testing

**Comment by sensecodevalue** [comment]
reat-ga를 확인해 봤습니다.
```
 spies.ga = jest.fn(); // 저희는 gtag등이 될 것 같습니다. 
 global.window.ga = spies.ga;
```
이런식으로 ga(gtag)를 window에 선언하고
```
 ReactGA.initialize('foo');
 ReactGA.ga('send', 'pageview', '/mypage');
 expect(spies.ga).toHaveBeenCalledTimes(2);
 expect(spies.ga).toHaveBeenNthCalledWith(1, 'create', 'foo', 'auto');
 expect(spies.ga).toHaveBeenNthCalledWith(2, 'send', 'pageview', '/mypage');
```
위의 방식으로  데이터를 mock으로 정확히 채워서 할 경우와, 아닐 경우 정도만 확인하는 것으로보아
이벤트 전송에 필요한 정보가 다들어 왔을 경우와 아닌 경우에 대해서만 테스트하고 정확하게 호출이되는지만 판단하면 좋을것 같습니다.

**Reply by josworks27** [reply]
@SenseCodeValue 
대윤님 덕분에 어느정도 감이 잡힌 것 같습니다 ! 감사합니다 ! 🙇🏻‍♂️

**Reply by greatSumini** [reply]
저도 동의합니다 ㅎㅎ
